### PR TITLE
Issue #4361: Implements the ClassAnalyzer and its tests.

### LINKED ISSUE #4361: Implements a ClassAnalyzer
This issue is from this comment from @keradus : https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4310#discussion_r271581322

Just like the FunctionsAnalyzer, CommentsAnalyzer and NamespacesAnalyzer, the class ClassAnalyzer will work with classes.

This has been created during the `#EUFOSSA Hackathon`.

### PR DESCRIPTION
**Pull request by drupol**
Fix #4361

### CODE REVIEW

**Comment by drupol** [comment]
I just rebased the branch against latest 2.12 branch.

**Comment by drupol** [comment]
This needs more work. I will finish it tonight.

**Comment by drupol** [comment]
This is now ready for another round of review.

@SpacePossum @keradus @dmvdbrugge @localheinz

File: tests/Tokenizer/Analyzer/ClassAnalyzerTest.php [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+ * @internal
+ *
+ * @covers \PhpCsFixer\Tokenizer\Analyzer\ClassAnalyzer
+ */
+final class ClassAnalyzerTest extends TestCase
+{
+    /**
+     * @param string $code
+     * @param bool   $hasExtends
+     * @param mixed  $classIndex
```
Comment (kubawerlos) [inline tests/Tokenizer/Analyzer/ClassAnalyzerTest.php]: Please fix the order and `mixed` type.

File: tests/Tokenizer/Analyzer/ClassAnalyzerTest.php [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+     *
+     * @dataProvider provideClassHasExtendsCases
+     */
+    public function testClassHasExtends($classIndex, $code, $hasExtends)
```
Comment (kubawerlos) [inline tests/Tokenizer/Analyzer/ClassAnalyzerTest.php]: I'd go either with testing here result of `getClassExtends` (so replacing `$hasExtends` with `null` or `TypeAnalysis` expectations) or add `hasClassExtends` (`isClassExtendsAnything`?) which would basically check if `getClassExtends` returns `null` or not.

Well - on the 2nd thought - I'd expect here we have direct tests for `getClassExtends` anyway.

File: src/Tokenizer/Analyzer/ClassAnalyzer.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ * @internal
+ */
+final class ClassAnalyzer
+{
+    /**
+     * @param Tokens $tokens
+     * @param int    $classIndex
+     *
+     * @return array
+     *               Contains information about the class being analyzed
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassAnalyzer.php]: What is the reason of having the description in separate line?

File: src/Tokenizer/Analyzer/ClassAnalyzer.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+     */
+    public function getClassDefinition(Tokens $tokens, $classIndex)
+    {
+        $openIndex = $tokens->getNextTokenOfKind($classIndex, ['{']);
+        $prev = $tokens->getPrevMeaningfulToken($classIndex);
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassAnalyzer.php]: `$prev` -> `$prevIndex`

File: tests/Tokenizer/Analyzer/ClassAnalyzerTest.php (hunk 1 above)
Comment (drupol) [reply tests/Tokenizer/Analyzer/ClassAnalyzerTest.php]: Fixed.

File: src/Tokenizer/Analyzer/ClassAnalyzer.php (hunk 3 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassAnalyzer.php]: Probably a mistake, I fixed it.

File: src/Tokenizer/Analyzer/ClassAnalyzer.php (hunk 4 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassAnalyzer.php]: Fixed.

File: tests/Tokenizer/Analyzer/ClassAnalyzerTest.php (hunk 2 above)
Comment (drupol) [reply tests/Tokenizer/Analyzer/ClassAnalyzerTest.php]: Okay, fixed now.

**Comment by kubawerlos** [comment]
In `Tokenizer/Analyzer` we have `ArgumentsAnalyzer`, `CommentsAnalyzer` and so on - I mean have plural names - therefore can we rename this analyzer to `ClassesAnalyzer`?

**Comment by drupol** [comment]
> In `Tokenizer/Analyzer` we have `ArgumentsAnalyzer`, `CommentsAnalyzer` and so on - I mean have plural names - therefore can we rename this analyzer to `ClassesAnalyzer`?

Fair enough. Done!

**Comment by SpacePossum** [comment]
@keradus can you look @ this one and https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4310 ?

**Comment by drupol** [comment]
Any news on this @SpacePossum

**Comment by SpacePossum** [comment]
please ping @keradus on this :)

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+            if (!$tokens[$classIndex]->isGivenKind(T_INTERFACE)) {
+                $implements = $tokens->findGivenKind(T_IMPLEMENTS, $classIndex, $openIndex);
+                $implements = \count($implements) ? $this->getClassInheritanceInfo($tokens, key($implements), 'numberOfImplements') : false;
+                $tokensAnalyzer = new TokensAnalyzer($tokens);
+                $anonymousClass = $tokensAnalyzer->isAnonymousClass($classIndex);
+            }
+        }
+
+        return [
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassesAnalyzer.php:51]: What about adding `<Name>Analysis` classes instead of associative arrays here and below in `getClassInheritanceInfo`? Would that be the next step (separate PR)?

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            'implements' => $implements,
+            'anonymousClass' => $anonymousClass,
+        ];
+    }
+
+    /**
+     * @param int $classIndex
+     *
+     * @return null|TypeAnalysis
+     *                           Get the class extends if any, null otherwise
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassesAnalyzer.php:65]: What happened here?

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+        return new TypeAnalysis($type, $typeStartIndex, $typeEndIndex);
+    }
+
+    /**
+     * @param int    $startIndex
+     * @param string $label
+     *
+     * @return array
+     *               Contains information about the class inheritance
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassesAnalyzer.php:98]: What happened here?

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+            $type .= $tokens[$typeEndIndex + 1]->getContent();
+        }
+
+        return new TypeAnalysis($type, $typeStartIndex, $typeEndIndex);
+    }
+
+    /**
+     * @param int    $startIndex
+     * @param string $label
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassesAnalyzer.php:95]: Can we have a description of what label is?

File: tests/Tokenizer/Analyzer/ClassesAnalyzerTest.php [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            static::assertSame($startIndex, $extend->getStartIndex());
+            static::assertSame($endIndex, $extend->getEndIndex());
+        } else {
+            static::assertNull($extend);
+        }
+    }
+
+    public function provideGetExtendsClassCases()
+    {
+        return [
```
Comment (kubawerlos) [inline tests/Tokenizer/Analyzer/ClassesAnalyzerTest.php:55]: Can we have test case(s) where `$classIndex` points to non-class, e.g. open tag and some `foreach` (that will have `{` as next token)?

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php (hunk 6 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php]: Fixed.

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php (hunk 7 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php]: Fixed.

File: tests/Tokenizer/Analyzer/ClassesAnalyzerTest.php (hunk 9 above)
Comment (drupol) [reply tests/Tokenizer/Analyzer/ClassesAnalyzerTest.php:55]: Fixed.

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php (hunk 8 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php]: Fixed.

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php (hunk 5 above)
Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php:51]: I don't get it, can you please explain a bit more ?

**Comment by drupol** [comment]
Dear @keradus, I adressed most of the points of your review.

Let me know what I should do to move this forward ;-)

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php (hunk 5 above)
Comment (kubawerlos) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php:51]: In some places (like [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.3/src/Tokenizer/Analyzer/NamespaceUsesAnalyzer.php#L39)) instead of returning array we return an object (like [this](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.3/src/Tokenizer/Analyzer/Analysis/NamespaceUseAnalysis.php)).

The returned objects are very simple and make using it easier than using an array.

I believe in the new code we want to use object approach over an array one.

Comment (keradus) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php:51]: Even more, the existing analysers should start returning objects as well.
- they are marked as internal anyway
- we can provide arrayaccess for compatibility reasons, if really needed

Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php:51]: Should we do that now or in a future PR ?

Comment (drupol) [reply src/Tokenizer/Analyzer/ClassesAnalyzer.php]: Ok I did it within this PR in commit https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4362/commits/889f98435d7d3f2251296220b936e4d0a1e96bbe. Let me know if it's ok now.

File: src/Fixer/ClassNotation/ClassDefinitionFixer.php [hunk 10]
```diff
@@ -153,43 +154,42 @@ protected function createConfigurationDefinition()
...
     {
-        $classDefInfo = $this->getClassyDefinitionInfo($tokens, $classyIndex);
+        $classAnalysis = $this->getClassyDefinitionInfo($tokens, $classyIndex);
 
         // PSR2 4.1 Lists of implements MAY be split across multiple lines, where each subsequent line is indented once.
         // When doing so, the first item in the list MUST be on the next line, and there MUST be only one interface per line.
-
-        if (false !== $classDefInfo['implements']) {
-            $classDefInfo['implements'] = $this->fixClassyDefinitionImplements(
+        if ([] !== $implements = $classAnalysis->getImplements()) {
```
Comment (kubawerlos) [inline src/Fixer/ClassNotation/ClassDefinitionFixer.php:161]: Can we extract the assignment before the `if`? Having it in the condition does not bring any value, but can lower readability.

File: src/Tokenizer/Analyzer/Analysis/ClassAnalysis.php [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * ClassAnalysis constructor.
+     *
+     * @param int  $start
+     * @param int  $class
+     * @param int  $open
+     * @param bool $anonymous
+     */
+    public function __construct($start, $class, $open, array $extends, array $implements, $anonymous)
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/Analysis/ClassAnalysis.php:58]: Can we have `@param` for array as well, with `int[]` type?

File: src/Tokenizer/Analyzer/Analysis/ClassAnalysis.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        $this->start = $start;
+        $this->class = $class;
+        $this->open = $open;
+        $this->extends = $extends;
+        $this->implements = $implements;
+        $this->anonymous = $anonymous;
+    }
+
+    /**
+     * @return mixed
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/Analysis/ClassAnalysis.php:69]: no `mixed` please, here and below.

File: src/Tokenizer/Analyzer/ClassesAnalyzer.php [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+        return new TypeAnalysis($type, $typeStartIndex, $typeEndIndex);
+    }
+
+    /**
+     * @param int    $startIndex
+     * @param string $label      an arbitrary name that will be used as key property
+     *                           in the return array
+     *
+     * @return array contains information about the class inheritance, keyed by
```
Comment (kubawerlos) [inline src/Tokenizer/Analyzer/ClassesAnalyzer.php:91]: The function above are already well updated to return objects, this one should too.

File: tests/Fixer/ClassNotation/ClassDefinitionFixerTest.php [hunk 14]
```diff
@@ -171,7 +171,7 @@ public function provideAnonymousClassesCases()
             ],
             [
                 "<?php \$a = new class('1a') implements A {};",
-                "<?php \$a = new class('1a')   implements\nA{};",
+                "<?php \$a = new class('1a')   implements A{};",
```
Comment (kubawerlos) [inline tests/Fixer/ClassNotation/ClassDefinitionFixerTest.php:174]: Why this change?
